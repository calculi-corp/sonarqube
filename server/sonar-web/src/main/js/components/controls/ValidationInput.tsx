/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import AlertErrorIcon from '../icons/AlertErrorIcon';
import AlertSuccessIcon from '../icons/AlertSuccessIcon';
import MandatoryFieldMarker from '../ui/MandatoryFieldMarker';
import HelpTooltip from './HelpTooltip';

export interface ValidationInputProps {
  description?: React.ReactNode;
  children: React.ReactNode;
  className?: string;
  error?: string;
  errorPlacement?: ValidationInputErrorPlacement;
  help?: string;
  id?: string;
  isInvalid: boolean;
  isValid: boolean;
  label?: React.ReactNode;
  required?: boolean;
}

export enum ValidationInputErrorPlacement {
  Right,
  Bottom
}

export default function ValidationInput(props: ValidationInputProps) {
  const {
    children,
    className,
    description,
    error,
    errorPlacement = ValidationInputErrorPlacement.Right,
    help,
    id,
    isInvalid,
    isValid,
    label,
    required
  } = props;
  const hasError = isInvalid && error !== undefined;

  let childrenWithStatus: React.ReactNode;
  if (errorPlacement === ValidationInputErrorPlacement.Right) {
    childrenWithStatus = (
      <>
        {children}
        {isValid && <AlertSuccessIcon className="spacer-left text-middle" />}
        {isInvalid && <AlertErrorIcon className="spacer-left text-middle" />}
        {hasError && <span className="little-spacer-left text-danger text-middle">{error}</span>}
      </>
    );
  } else {
    childrenWithStatus = (
      <>
        {children}
        {isValid && <AlertSuccessIcon className="spacer-left text-middle" />}
        <div className="spacer-top">
          {isInvalid && <AlertErrorIcon className="text-middle" />}
          {hasError && <span className="little-spacer-left text-danger text-middle">{error}</span>}
        </div>
      </>
    );
  }

  return (
    <div className={className}>
      {id && label && (
        <label htmlFor={id}>
          <span className="text-middle">
            <strong>{label}</strong>
            {required && <MandatoryFieldMarker />}
          </span>
          {help && <HelpTooltip className="spacer-left" overlay={help} />}
        </label>
      )}
      <div className="little-spacer-top spacer-bottom">{childrenWithStatus}</div>
      {description && <div className="note abs-width-400">{description}</div>}
    </div>
  );
}
