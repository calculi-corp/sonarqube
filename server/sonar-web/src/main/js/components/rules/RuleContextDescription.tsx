/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { translate } from '../../helpers/l10n';
import RadioToggle from '../controls/RadioToggle';
import { sanitizeString } from '../../helpers/sanitize';
import { RuleDescriptionSection } from '../../apps/coding-rules/rule';
import OtherContextOption from './OtherContextOption';

const OTHERS_KEY = 'others';

interface Props {
  description: RuleDescriptionSection[];
}

interface State {
  contexts: RuleDescriptionContextDisplay[];
  selectedContext: RuleDescriptionContextDisplay;
}

interface RuleDescriptionContextDisplay {
  displayName: string;
  content: string;
  key: string;
}

export default class RuleContextDescription extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = this.computeState(props.description);
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.description !== this.props.description) {
      this.setState(this.computeState(this.props.description));
    }
  }

  computeState = (descriptions: RuleDescriptionSection[]) => {
    const contexts = descriptions
      .map(sec => ({
        displayName: sec.context?.displayName || '',
        content: sec.content,
        key: sec.key.toString()
      }))
      .filter(sec => sec.displayName !== '')
      .sort((a, b) => a.displayName.localeCompare(b.displayName));

    if (contexts.length > 0) {
      contexts.push({
        displayName: translate('coding_rules.description_context_other'),
        content: '',
        key: OTHERS_KEY
      });
    }

    return {
      contexts,
      selectedContext: contexts[0]
    };
  };

  handleToggleContext = (value: string) => {
    const { contexts } = this.state;

    const selected = contexts.find(ctxt => ctxt.displayName === value);
    if (selected) {
      this.setState({ selectedContext: selected });
    }
  };

  render() {
    const { contexts } = this.state;
    const { selectedContext } = this.state;

    const options = contexts.map(ctxt => ({
      label: ctxt.displayName,
      value: ctxt.displayName
    }));

    return (
      <div className="rules-context-description">
        <h2 className="rule-contexts-title">
          {translate('coding_rules.description_context_title')}
        </h2>
        <RadioToggle
          className="big-spacer-bottom"
          name="filter"
          onCheck={this.handleToggleContext}
          options={options}
          value={selectedContext.displayName}
        />
        {selectedContext.key === OTHERS_KEY ? (
          <OtherContextOption />
        ) : (
          <div
            /* eslint-disable-next-line react/no-danger */
            dangerouslySetInnerHTML={{ __html: sanitizeString(selectedContext.content) }}
          />
        )}
      </div>
    );
  }
}
