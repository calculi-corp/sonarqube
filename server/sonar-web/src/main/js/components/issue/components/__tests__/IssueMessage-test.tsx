/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { RuleStatus } from '../../../../types/rules';
import { ButtonLink } from '../../../controls/buttons';
import IssueMessage, { IssueMessageProps } from '../IssueMessage';

jest.mock('react', () => {
  return {
    ...jest.requireActual('react'),
    useContext: jest
      .fn()
      .mockImplementation(() => ({ externalRulesRepoNames: {}, openRule: jest.fn() }))
  };
});

it('should render correctly', () => {
  expect(shallowRender()).toMatchSnapshot('default');
  expect(shallowRender({ engine: 'js' })).toMatchSnapshot('with engine info');
  expect(shallowRender({ quickFixAvailable: true })).toMatchSnapshot('with quick fix');
  expect(shallowRender({ manualVulnerability: true })).toMatchSnapshot('is manual vulnerability');
  expect(shallowRender({ ruleStatus: RuleStatus.Deprecated })).toMatchSnapshot(
    'is deprecated rule'
  );
  expect(shallowRender({ ruleStatus: RuleStatus.Removed })).toMatchSnapshot('is removed rule');
  expect(shallowRender({ displayWhyIsThisAnIssue: false })).toMatchSnapshot(
    'hide why is it an issue'
  );
});

it('should open why is this an issue workspace', () => {
  const openRule = jest.fn();
  (React.useContext as jest.Mock).mockImplementationOnce(() => ({
    externalRulesRepoNames: {},
    openRule
  }));
  const wrapper = shallowRender();
  wrapper.find(ButtonLink).simulate('click');

  expect(openRule).toBeCalled();
});

function shallowRender(props: Partial<IssueMessageProps> = {}) {
  return shallow<IssueMessageProps>(
    <IssueMessage
      manualVulnerability={false}
      message="Reduce the number of conditional operators (4) used in the expression"
      displayWhyIsThisAnIssue={true}
      ruleKey="javascript:S1067"
      {...props}
    />
  );
}
