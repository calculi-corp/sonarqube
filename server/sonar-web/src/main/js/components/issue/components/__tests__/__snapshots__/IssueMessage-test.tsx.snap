// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly: default 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: hide why is it an issue 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
  </div>
</Fragment>
`;

exports[`should render correctly: is deprecated rule 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
    <DocumentationTooltip
      className="spacer-left"
      content="rules.status.DEPRECATED.help"
      links={
        Array [
          Object {
            "href": "/documentation/user-guide/rules/",
            "label": "see_x.rules",
          },
        ]
      }
    >
      <span
        className="spacer-right badge badge-error"
      >
        issue.resolution.DEPRECATED.rule_deprecated
      </span>
    </DocumentationTooltip>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: is manual vulnerability 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
    <Tooltip
      overlay="issue.manual_vulnerability.description"
    >
      <div
        className="badge spacer-right text-baseline"
      >
        issue.manual_vulnerability
      </div>
    </Tooltip>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: is removed rule 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
    <DocumentationTooltip
      className="spacer-left"
      content="rules.status.REMOVED.help"
      links={
        Array [
          Object {
            "href": "/documentation/user-guide/rules/",
            "label": "see_x.rules",
          },
        ]
      }
    >
      <span
        className="spacer-right badge badge-error"
      >
        issue.resolution.REMOVED.rule_removed
      </span>
    </DocumentationTooltip>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: with engine info 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
    <Tooltip
      overlay="issue.from_external_rule_engine.js"
    >
      <div
        className="badge spacer-right text-baseline"
      >
        js
      </div>
    </Tooltip>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;

exports[`should render correctly: with quick fix 1`] = `
<Fragment>
  <div
    className="display-inline-flex-center issue-message break-word"
  >
    <span
      className="spacer-right"
    >
      Reduce the number of conditional operators (4) used in the expression
    </span>
    <Tooltip
      mouseLeaveDelay={0.5}
      overlay={
        <FormattedMessage
          defaultMessage="issue.quick_fix_available_with_sonarlint"
          id="issue.quick_fix_available_with_sonarlint"
          values={
            Object {
              "link": <a
                href="https://www.sonarqube.org/sonarlint/?referrer=sonarqube-quick-fix"
                rel="noopener noreferrer"
                target="_blank"
              >
                SonarLint
              </a>,
            }
          }
        />
      }
    >
      <SonarLintIcon
        className="it__issues-sonarlint-quick-fix spacer-right"
        size={15}
      />
    </Tooltip>
  </div>
  <ButtonLink
    aria-label="issue.why_this_issue.long"
    className="issue-see-rule spacer-right text-baseline"
    onClick={[Function]}
  >
    issue.why_this_issue
  </ButtonLink>
</Fragment>
`;
